18-websockets/classworks/Gavruk_Alexander/src/pages/Chat/Chat.jsx

```diff
@@ -0,0 +1,129 @@
+import React, { useEffect, useState } from "react";
+import socketIOClient from "socket.io-client";
+
+import './Chat.css';
+import Message from './components/Message';
+
+const ENDPOINT = "http://localhost:3001/";
+const socket = socketIOClient(ENDPOINT);
+
+function Chat() {
+  const [usernameInputValue, setUsernameInputValue] = useState('');
+  const [msgInputValue, setMsgInputValue] = useState('');
+  const [username, setUsername] = useState('');
+  const [messages, setMessages] = useState([]);
+  const [typingUsers, setTypingUsers] = useState([]);
+  const [timerId, setTimerId] = useState(null);
+
+  const onUsernameSubmit = (username) => {
+    socket.emit('set_username', username);
+    setUsernameInputValue('');
+  }
+
+  const onSendMessage = (msg) => {
+    socket.emit('new_message', { username, value: msg });
+    setMsgInputValue('');
+  }
+
+  const onType = (username) => {
+    const timerId = setTimeout(() => {
+      setTypingUsers((typingUsers) => typingUsers.filter((name) => name !== username))
+    }, 7000);
+    setTimerId(timerId);
+
+    setTypingUsers((prevTypings) => {
+      const currentTypings = prevTypings || typingUsers;
+      if (currentTypings.findIndex((name) => name === username) === -1) {
+        return [...currentTypings, username];
+      } else {
+        return prevTypings;
+      }
+    })
+  }
+
+  const onInputChange = (value) => {
+    setMsgInputValue(value);
+    typing();
```

Comment:
if the user is a super typer man, you will get a huge number of requests from him. A good practice is to use the debounce method for calling function no more than one time per 2-4 secs.
Unfortunately, javascript doesn't have this method. You can write it by yourself or use some library.